ingest/ledgerbackend/cloud_storage_backend.go

```diff
@@ -0,0 +1,210 @@
+
+	defer reader.Close()
+
+	gzipReader, err := gzip.NewReader(reader)
+	if err != nil {
+		return xdr.LedgerCloseMeta{}, errors.Wrapf(err, "failed getting file: %s", objectKey)
+	}
+
+	defer gzipReader.Close()
+
+	objectBytes, err := io.ReadAll(gzipReader)
+	if err != nil {
+		return xdr.LedgerCloseMeta{}, errors.Wrapf(err, "failed reading file: %s", objectKey)
+	}
+
+	err = ledgerCloseMetaBatch.UnmarshalBinary(objectBytes)
+	if err != nil {
+		return xdr.LedgerCloseMeta{}, errors.Wrapf(err, "failed unmarshalling file: %s", objectKey)
+	}
+
+	ledgerCloseMetasIndex := sequence - uint32(ledgerCloseMetaBatch.StartSequence)
+	ledgerCloseMeta := ledgerCloseMetaBatch.LedgerCloseMetas[ledgerCloseMetasIndex]
+
+	return ledgerCloseMeta, nil
+}
+
+// PrepareRange checks if the starting and ending (if bounded) ledgers exist.
+func (csb *CloudStorageBackend) PrepareRange(ctx context.Context, ledgerRange Range) error {
+	_, err := csb.GetLedger(ctx, ledgerRange.from)
+	if err != nil {
+		return errors.Wrapf(err, "error getting ledger %d", ledgerRange.from)
+	}
+
+	if ledgerRange.bounded {
+		_, err := csb.GetLedger(ctx, ledgerRange.to)
+		if err != nil {
+			return errors.Wrapf(err, "error getting ending ledger %d", ledgerRange.to)
+		}
+	}
+
+	return nil
+}
+
+// IsPrepared is a no-op for CloudStorageBackend.
+func (csb *CloudStorageBackend) IsPrepared(ctx context.Context, ledgerRange Range) (bool, error) {
+	return true, nil
+}
+
+// Close is a no-op for CloudStorageBackend.
+func (csb *CloudStorageBackend) Close() error {
+	return nil
+}
+
+// TODO: Should this function also be modified and added to support/datastore?
+// This function should be shared between ledger exporter and this legerbackend reader
+func GetObjectKeyFromSequenceNumber(ledgerSeq uint32, ledgersPerFile uint32, filesPerPartition uint32) (string, error) {
+	var objectKey string
+
+	if ledgersPerFile < 1 {
+		return "", errors.Errorf("Invalid ledgers per file (%d): must be at least 1", ledgersPerFile)
+	}
+
+	if filesPerPartition > 1 {
+		partitionSize := ledgersPerFile * filesPerPartition
+		partitionStart := (ledgerSeq / partitionSize) * partitionSize
+		partitionEnd := partitionStart + partitionSize - 1
+		objectKey = fmt.Sprintf("%d-%d/", partitionStart, partitionEnd)
+	}
+
+	fileStart := (ledgerSeq / ledgersPerFile) * ledgersPerFile
+	fileEnd := fileStart + ledgersPerFile - 1
+	objectKey += fmt.Sprintf("%d", fileStart)
+
+	// Multiple ledgers per file
+	if fileStart != fileEnd {
+		objectKey += fmt.Sprintf("-%d", fileEnd)
+	}
+	objectKey += fileSuffix
+
+	return objectKey, nil
+}
+
+func getLatestDirectory(directories []string) (string, error) {
+	var latestDirectory string
+	largestDirectoryLedger := 0
+
+	for _, dir := range directories {
+		// dir follows the format of "ledgers/<network>/<start>-<end>"
+		// Need to split the dir string to retrieve the <end> ledger value to get the latest directory
+		dirParts := strings.Split(dir, "/")
+		lastDirPart := dirParts[len(dirParts)-1]
```

Comment:
path-specific code would be slightly cleaner/safer:

```suggestion
		// dir follows the format of "ledgers/<network>/<start>-<end>"
		// Need to split the dir string to retrieve the <end> ledger value to get the latest directory
		_, lastDirPart := path.Split(dir)
```

also, any chance `dir` has a trailing slash? as that would break both of these approaches by making `lastDirPart` empty, maybe add a [`strings.TrimSuffix(dir, "/")`](https://pkg.go.dev/strings#TrimSuffix) to be safe?

Reply:
I don't think it would have a trailing slash but it doesn't hurt to add the check in